LE Audio: CAP initiator unicast procedures

Thalley opened the pull request:
Implements the CAP initiator unicast audio procedures

Depends on https://github.com/zephyrproject-rtos/zephyr/pull/47385
fixes https://github.com/zephyrproject-rtos/zephyr/issues/47311

Thalley committed:
Bluetooth: Audio: Implement CAP unicast audio start

Implement the CAP unicast start procedure, which
will put one or more CAP streams from the idle state
to the streaming state, across or more ACL connections.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: Add support for CAP unicast start in shell

This modifies the audio.c shell module to support
multiple connections.

This adds support for the cap_initiator unicast-start command,
which supports multiple endpoints on multiple devices.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
tests: Bluetooth: Audio: Add CAP unicast start BSIM test

The test only works partially until the controller and
babblesim supports ISO, as it needs that in order to
perform the unicast start procedure.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: Implement CAP unicast metadata update

Implement the CAP unicast metadata update procedure.

The procedure lets an application update the metadata of one
or more CAP streams in a single function call, while also ensuring
that the provided metadata follows the requirement from CAP.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: Add CAP unicast update shell command

Add shell command for the CAP unicast update procedure.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: Implement CAP unicast stop procedure

Implements the CAP unicast stop procedures, which is used
to stop (either disable or release) multiple streams.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: Add CAP unicast stop shell command

Add shell command for the CAP unicast stop procedure.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Tests: Bluetooth: Add CAP unicast update test

Add test to run the CAP initiator unicast update API.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Tests: Bluetooth: Add unicast stop to CAP tests

Add test of the unicast stop procedure for the CAP
initiator.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: CAP Initiator handle unexpected ASE changes

When performing the CAP Unicast Start procedure, the remote
server(s) may change the state of the ASE before the procedure
is complete.

If the state change is unexpected, we stop the procedure and
call the callback, and then the caller can decide what to
do after that (presumably call the stop procedure for the
streams that may be setup).

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
doc: Bluetooth: Add CAP documentation

Add documentation for CAP by adding the reference to the API
document.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: Shell: Use same streams and group for BAP and CAP

Use the same unicast streams and group for the BAP and CAP
shell commands. This results in maximum flexibility, as well
as the niceness of use from CAP, as well as less memory usage.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

MariuszSkamra commented on subsys/bluetooth/audio/cap_initiator.c:
```diff
@@ -378,22 +476,868 @@ int bt_cap_initiator_unicast_discover(struct bt_conn *conn)
+	 * use the reference from the first stream
+	 */
+	unicast_group = active_procedure.streams[0]->bap_stream.unicast_group;
+	active_procedure.stream_done = 0U;
+	active_procedure.stream_initiated = 0U;
+
+	active_procedure.subproc_type = CAP_UNICAST_SUBPROCEDURE_TYPE_QOS_CONFIG;
```
```suggestion
static void cap_set_active_procedure(procedure)
{
	active_procedure.stream_done = 0U;
	active_procedure.stream_initiated = 0U;
	active_procedure.subproc_type = CAP_UNICAST_SUBPROCEDURE_TYPE_QOS_CONFIG;
}
```

`cap_set_active_procedure(CAP_UNICAST_SUBPROCEDURE_TYPE_QOS_CONFIG);`

MariuszSkamra commented on subsys/bluetooth/audio/cap_initiator.c:
```diff
@@ -378,22 +476,868 @@ int bt_cap_initiator_unicast_discover(struct bt_conn *conn)
+		return;
+	}
+
+	active_procedure.stream_done = 0U;
+	active_procedure.stream_initiated = 0U;
+
+	active_procedure.subproc_type = CAP_UNICAST_SUBPROCEDURE_TYPE_ENABLE;
```
ditto

MariuszSkamra commented on subsys/bluetooth/audio/cap_initiator.c:
```diff
@@ -378,22 +476,868 @@ int bt_cap_initiator_unicast_discover(struct bt_conn *conn)
+		return;
+	}
+
+	active_procedure.stream_done = 0U;
+	active_procedure.stream_initiated = 0U;
+
+	active_procedure.subproc_type = CAP_UNICAST_SUBPROCEDURE_TYPE_START;
```
ditto

MariuszSkamra commented on subsys/bluetooth/audio/cap_initiator.c:
```diff
@@ -378,22 +476,868 @@ int bt_cap_initiator_unicast_discover(struct bt_conn *conn)
+		}
+	}
+
+	atomic_set_bit(active_procedure.proc_state_flags,
+		       CAP_UNICAST_PROCEDURE_STATE_ACTIVE);
+	active_procedure.stream_cnt = count;
+
+	active_procedure.subproc_type = CAP_UNICAST_SUBPROCEDURE_TYPE_META_UPDATE;
```
ditto

MariuszSkamra commented on subsys/bluetooth/audio/cap_initiator.c:
```diff
@@ -378,22 +476,868 @@ int bt_cap_initiator_unicast_discover(struct bt_conn *conn)
+		return -EALREADY;
+	}
+
+	atomic_set_bit(active_procedure.proc_state_flags,
+		       CAP_UNICAST_PROCEDURE_STATE_ACTIVE);
+	active_procedure.stream_cnt = stream_cnt;
+	active_procedure.unicast_group = unicast_group;
+
+	active_procedure.subproc_type = CAP_UNICAST_SUBPROCEDURE_TYPE_RELEASE;
```
ditto

MariuszSkamra commented on subsys/bluetooth/audio/cap_initiator.c:
```diff
@@ -378,22 +476,868 @@ int bt_cap_initiator_unicast_discover(struct bt_conn *conn)
+	/* Store the information about the active procedure so that we can
+	 * continue the procedure after each step
+	 */
+	atomic_set_bit(active_procedure.proc_state_flags,
+		       CAP_UNICAST_PROCEDURE_STATE_ACTIVE);
+	active_procedure.stream_cnt = param->count;
+
+	active_procedure.subproc_type = CAP_UNICAST_SUBPROCEDURE_TYPE_CODEC_CONFIG;
```
ditto

MariuszSkamra commented on subsys/bluetooth/audio/cap_initiator.c:
```diff
@@ -378,22 +476,868 @@ int bt_cap_initiator_unicast_discover(struct bt_conn *conn)
+			cap_stream, active_procedure.stream_done,
+			active_procedure.stream_cnt);
+
+		if (active_procedure.stream_done < active_procedure.stream_cnt) {
```
```suggestion
static bool cap_procedure_is_done(void)
{
return active_procedure.stream_done == active_procedure.stream_cnt;
}
```
use some kind of helper function to be reused in other places places

MariuszSkamra commented on subsys/bluetooth/audio/cap_initiator.c:
```diff
@@ -378,22 +476,868 @@ int bt_cap_initiator_unicast_discover(struct bt_conn *conn)
+			cap_stream, active_procedure.stream_done,
+			active_procedure.stream_cnt);
+
+		if (active_procedure.stream_done < active_procedure.stream_cnt) {
```
here

MariuszSkamra commented on subsys/bluetooth/audio/cap_initiator.c:
```diff
@@ -378,22 +476,868 @@ int bt_cap_initiator_unicast_discover(struct bt_conn *conn)
+			cap_stream, active_procedure.stream_done,
+			active_procedure.stream_cnt);
+
+		if (active_procedure.stream_done < active_procedure.stream_cnt) {
```
here

MariuszSkamra commented on subsys/bluetooth/audio/cap_initiator.c:
```diff
@@ -378,22 +476,868 @@ int bt_cap_initiator_unicast_discover(struct bt_conn *conn)
+	 * only do this one by one due to a restriction in the ISO layer
+	 * (maximum 1 outstanding ISO connection request at any one time).
+	 */
+	if (active_procedure.stream_done < active_procedure.stream_cnt) {
```
here

MariuszSkamra commented on subsys/bluetooth/audio/cap_initiator.c:
```diff
@@ -378,22 +476,868 @@ int bt_cap_initiator_unicast_discover(struct bt_conn *conn)
+			active_procedure.stream_cnt);
+	}
+
+	if (active_procedure.stream_done < active_procedure.stream_cnt) {
```
here

MariuszSkamra commented on subsys/bluetooth/audio/cap_initiator.c:
```diff
@@ -378,22 +476,868 @@ int bt_cap_initiator_unicast_discover(struct bt_conn *conn)
+			active_procedure.stream_cnt);
+	}
+
+	if (active_procedure.stream_done < active_procedure.stream_cnt) {
```
here

MariuszSkamra reviewed (changes requested):
My review is incomplete yet. For now I left few comments you may want to address

Thalley commented:
Rebased to solve merge conflicts

Thalley commented on subsys/bluetooth/audio/cap_initiator.c:
```diff
@@ -196,15 +205,98 @@ int bt_cap_initiator_broadcast_get_base(struct bt_cap_broadcast_source *source,
 
 #if defined(CONFIG_BT_AUDIO_UNICAST_CLIENT)
-static const struct bt_uuid *cas_uuid = BT_UUID_CAS;
+enum {
+	CAP_UNICAST_PROCEDURE_STATE_ACTIVE,
+	CAP_UNICAST_PROCEDURE_STATE_ABORTED,
+
+	CAP_UNICAST_PROCEDURE_STATE_FLAG_NUM,
+} cap_unicast_procedure_state;
```
Hmm, not at the moment. Used flags to use the atomic functions. The alternative would be to guard the access to the state via another locking mechanism, as this can be set and read by the application thread or the BT RX thread

Thalley replied:
Actually yes: A procedure may be active and aborted. If we are awaiting responses from other devices after one have failed, the procedure is still active in the sense that we cannot start any more, and it is aborted to stop the stack from going forwards with the next step. 

We could introduce a new state, IDLE, and check for that instead of `ACTIVE` if we want to keep this a single value.

Thalley committed:
Bluetooth: Audio: CAP initiator streamline procedure states

Add new helper functions to increase readability and to
ensure correctness when aborting procedures.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley commented:
@MariuszSkamra Thanks for the excellent suggestions. I concluded that it would be much easier for me to handle the suggestions in a new commit, so added a new commit to streamline the procedure checks.

Thalley commented:
Rebased after the BAP API rename

larsgk commented on subsys/bluetooth/audio/cap_initiator.c:
```diff
@@ -190,15 +199,141 @@ int bt_cap_initiator_broadcast_get_base(struct bt_cap_broadcast_source *source,
+
+struct cap_unicast_proc {
+	ATOMIC_DEFINE(proc_state_flags, CAP_UNICAST_PROC_STATE_FLAG_NUM);
+	/* Total number of streams in the procedure*/
+	size_t stream_cnt;
+	/* Number of streams where a subprocedure have been started */
+	size_t stream_initiated;
+	/* Number of streams done with the procedure */
+	size_t stream_done;
```
Would it make sense to postfix `stream_initiated` and `stream_done` with `_cnt` to more clearly indicate that they are counters and will be compared with `stream_cnt` (~ logically connected)

larsgk commented on subsys/bluetooth/shell/cap_initiator.c:
```diff
@@ -79,10 +474,23 @@ static int cmd_cap_initiator(const struct shell *sh, size_t argc, char **argv)
-					     cmd_cap_initiator_discover, 1, 0),
+	SHELL_CMD_ARG(discover, NULL,
+		      "Discover CAS", cmd_cap_initiator_discover, 1, 0),
+	SHELL_CMD_ARG(unicast-start, NULL,
```
is it possible to use stop/start_sine from the bap shell on multiple coordinated set devices? - also, if so, maybe we should be able to set the frequency to distinguish different channels

Thalley replied:
Not currently. The current implementation of the `start_sine` command only support a single TXing stream.

MariuszSkamra commented on subsys/bluetooth/audio/cap_stream.c:
```diff
@@ -15,6 +21,12 @@ static void cap_stream_configured_cb(struct bt_bap_stream *bap_stream,
 							bap_stream);
 	struct bt_bap_stream_ops *ops = cap_stream->ops;
 
+	LOG_DBG("%p", cap_stream);
+
+	if (IS_ENABLED(CONFIG_BT_CAP_INITIATOR)) {
+		bt_cap_initiator_codec_configured(cap_stream);
+	}
+
 	if (ops != NULL && ops->configured != NULL) {
 		ops->configured(bap_stream, pref);
 	}
```
It looks like dependency is mixed here. CAP initiator code calls `bt_cap_stream_ops_register_bap` which registers this callback and from this callback, the `bt_cap_initiator_codec_configured` is called.
Probably, the cap_initiator.c shall have it's own bt_bap_stream_ops.

Thalley replied:
> Probably, the cap_initiator.c shall have it's own bt_bap_stream_ops.

Yes, this code is from before the BAP rename, where the callback structure was called `bt_audio_stream_ops` and was more generic. 

CAP is effectively on top of BAP, so the BAP stream ops are registered by CAP, and the CAP stream parameters are registered by the user of CAP. 

So CAP registers a set of callbacks for the BAP stream, and uses those callbacks to execute the CAP procedures, while also calling the CAP stream callbacks registered by the user of CAP. 

I am not sure I understand the issue you may be on to here, so please elaborate a bit more :)

MariuszSkamra replied:
Ah, I thought it's on BAP level (`bap_stream.c`), meaning `bt_cap_initiator_codec_configured` is called from BAP level. If it's `cap_stream.c` it's OK :D

Thalley commented on subsys/bluetooth/audio/cap_initiator.c:
```diff
@@ -372,21 +513,814 @@ int bt_cap_initiator_unicast_discover(struct bt_conn *conn)
+		return false;
+	}
+
+	/* Only initiate the release procedure for streams not already idle */
```
I guess the code is self-explanatory here :D

Thalley commented:
Test is failing due to https://github.com/zephyrproject-rtos/zephyr/pull/55936

Thalley commented:
Rebased to solve merge conflicts

Thalley commented:
Rebased due to BSIM structure changes
